### ISSUE TITLE
The sourcemaps bug is still extant

### ISSUE THREAD

**Issue by djalbat** [swc-project/swc#4112]
### Describe the bug

The sourcemaps are breaking both Mozilla's `source-map` package and `esbuild`. I reported this issue previously:

https://github.com/swc-project/swc/issues/3772

I have found a simpler repository to test. Please try this one:

https://github.com/djalbat/easy-richtextarea

Clone this repository then run the usual `npm install` followed by `npm run build-debug` and you will see the errors. Alternatively, try the input code below and test with the `source-map` package.

### Input code

```typescript
"use strict";

import { React, window } from "easy";

window.assign({
  React
});
```

### Config

```json
{
    "jsc": {
        "parser": {
            "syntax": "ecmascript",
            "jsx": true
        }
    },
    "module": {
        "type": "commonjs"
    }
}
```

### Expected behavior

There should be no errors produced when the transpiled code is passed to bundlers or the `source-map` package.

### Actual behavior

This is the output:

```
"use strict";
var _easy = require("easy");
_easy.window.assign({
    React: _easy.React
});

//# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbIi4uLy4uL3NyYy9leGFtcGxlL3ByZWFtYmxlLmpzIl0sInNvdXJjZXNDb250ZW50IjpbIlwidXNlIHN0cmljdFwiO1xuXG5pbXBvcnQgeyBSZWFjdCwgd2luZG93IH0gZnJvbSBcImVhc3lcIjtcblxud2luZG93LmFzc2lnbih7XG4gIFJlYWN0XG59KTtcbiJdLCJuYW1lcyI6WyJ3aW5kb3ciLCJhc3NpZ24iLCJSZWFjdCJdLCJtYXBwaW5ncyI6IkFBQUEsWUFBWSxDQUFDO0FBRWlCLElBQUEsS0FBTSxXQUFOLE1BQU0sQ0FBQTtBQUVwQ0EsS0FBTSxRQUFDQyxNQUFNLENBQUM7SUNKZCxBREtFQyxLQUFLLEVBQUxBLEtBQUssTUFBTEE7Q0FDRCxDQUFDLENBQUMifQ==
```

It results in the following error from `esbuild`:

```
 lib/example/preamble.js#sourceMappingURL:1:295: warning: Bad "mappings" data in source map at character 75: Invalid source index value: 1
    1 │ ... React, window } from \"easy\";\n\nwindow.assign({\n  React\n});\n"],"names":["window","assign","React"],"mappings":"AAAA,YAAY,CAAC;AAEiB,IAAA,KAAM,WAAN,MAAM,CAAA;AAEpCA,KAAM,QAACC,MAAM,CAAC;ICJd,ADKEC,KAAK,EAALA,KAAK,MAALA;CACD,CAAC,CAAC"}
      ╵                                                                                                                                                                                                    ^
```

The `source-map` package produces a similar error.

### Version

1.2.159

**Comment by kwonoj** [comment]
Please provide reproducible code does not involve 3rd party tooling, specifically, in here `watchful-cli`. I do not know details what those tool actually does, however peeking through stacks it augments generated sourcemap from swc's output https://github.com/djalbat/watchful-cli/blob/521d4bbccc1c2c107e7d9e8ed665b9a69a4fe5eb/bin/utilities/transpile.js#L111-L132

```
            const { code, map } = output,
                  mapJSON = JSON.parse(map),
                  sourceFileName = sourceFileNameFromSourceFilePathAndTargetFilePath(sourceFilePath, targetFilePath),
                  source = sourceFileName,  ///
                  sources = [
                    source
                  ],
                  sourceContent = sourceFileContent,  ///
                  sourcesContent = [
                    sourceContent
                  ];

            Object.assign(mapJSON, {
              sources,
              sourcesContent
            });

            const mapJSONString = JSON.stringify(mapJSON),
                  base64EncodedMapJSONString = Buffer.from(mapJSONString).toString(BASE64_ENCODING);

            targetFileContent = `${code}
${SOURCE_MAP_PREAMBLE}${base64EncodedMapJSONString}`; ///
```

and swc's options like `sourceMaps: inline` is ignored. We need isolated reproducible case from raw output of swc itself, not the output augmented by other tools.

**Comment by RiESAEX** [comment]
I cloned it and run `swc ./src -d ./lib -s inline`, and [source-map-visualization](http://sokra.github.io/source-map-visualization/) parsed `lib/example/preamble.js` successfully

**Comment by djalbat** [comment]
I wondered whether inserting the source code manually, so to speak, was at the root of the issue. I will take a look this evening. Thank you for your input.

**Comment by djalbat** [comment]
@[RiESAEX, which versions are you using? This morning I installed:

```
+ @swc/core@1.2.160
+ @swc/cli@0.1.56
```
Running `swc ./src -d ./lib -s inline` does not include the source, however.

**Comment by djalbat** [comment]
I have done some more research on this and I am still not sure what the root cause is. Try the following:

```
"use strict";
const { React, window } = require("easy");
window.assign({
    React: React
});
```

Note the `require()` statement here. This produces the following output:

```
"use strict";
var ref = require("easy"), React = ref.React, window = ref.window;
window.assign({
    React: React
});

//# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbIi4uLy4uL3NyYy9leGFtcGxlL3ByZWFtYmxlLmpzIl0sInNvdXJjZXNDb250ZW50IjpbIlwidXNlIHN0cmljdFwiO1xuY29uc3QgeyBSZWFjdCwgd2luZG93IH0gPSByZXF1aXJlKFwiZWFzeVwiKTtcbndpbmRvdy5hc3NpZ24oe1xuICAgIFJlYWN0OiBSZWFjdFxufSk7XG4iXSwibmFtZXMiOlsicmVxdWlyZSIsIlJlYWN0Iiwid2luZG93IiwiYXNzaWduIl0sIm1hcHBpbmdzIjoiQUFBQSxZQUFZLENBQUM7QUFDYixJQUEwQkEsR0FBZSxHQUFmQSxPQUFPLENBQUMsTUFBTSxDQUFDLEVBQWpDQyxLQUFLLEdBQWFELEdBQWUsQ0FBakNDLEtBQUssRUFBRUMsTUFBTSxHQUFLRixHQUFlLENBQTFCRSxNQUFNLEFBQXFCO0FBQzFDQSxNQUFNLENBQUNDLE1BQU0sQ0FBQztJQUNWRixLQUFLLEVBQUVBLEtBQUs7Q0FDZixDQUFDLENBQUMifQ==
```

The source map is:

```
{
   "version":3,
   "sources":[
      "../../src/example/preamble.js"
   ],
   "sourcesContent":[
      "\"use strict\";\nconst { React, window } = require(\"easy\");\nwindow.assign({\n    React: React\n});\n"
   ],
   "names":[
      "require",
      "React",
      "window",
      "assign"
   ],
   "mappings":"AAAA,YAAY,CAAC;AACb,IAA0BA,GAAe,GAAfA,OAAO,CAAC,MAAM,CAAC,EAAjCC,KAAK,GAAaD,GAAe,CAAjCC,KAAK,EAAEC,MAAM,GAAKF,GAAe,CAA1BE,MAAM,AAAqB;AAC1CA,MAAM,CAACC,MAAM,CAAC;IACVF,KAAK,EAAEA,KAAK;CACf,CAAC,CAAC"
}
```

In this case, everything is fine.

Now try the following:

```
"use strict";
import { React, window } from "easy";
window.assign({
    React: React
});
```
Note the more usual `import()` statement here. This produces the following output:

```
"use strict";
var _easy = require("easy");
_easy.window.assign({
    React: _easy.React
});

//# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbIi4uLy4uL3NyYy9leGFtcGxlL3ByZWFtYmxlLmpzIl0sInNvdXJjZXNDb250ZW50IjpbIlwidXNlIHN0cmljdFwiO1xuaW1wb3J0IHsgUmVhY3QsIHdpbmRvdyB9IGZyb20gXCJlYXN5XCI7XG53aW5kb3cuYXNzaWduKHtcbiAgICBSZWFjdDogUmVhY3Rcbn0pO1xuIl0sIm5hbWVzIjpbIndpbmRvdyIsImFzc2lnbiIsIlJlYWN0Il0sIm1hcHBpbmdzIjoiQUFBQSxZQUFZLENBQUM7QUFDaUIsSUFBQSxLQUFNLFdBQU4sTUFBTSxDQUFBO0FBQ3BDQSxLQUFNLFFBQUNDLE1BQU0sQ0FBQztJQ0ZkLEFER0lDLEtBQUssRUFBRUEsS0FBSyxNQUFaQTtDQUNILENBQUMsQ0FBQyJ9
```
The source map is:

```
{
   "version":3,
   "sources":[
      "../../src/example/preamble.js"
   ],
   "sourcesContent":[
      "\"use strict\";\nimport { React, window } from \"easy\";\nwindow.assign({\n    React: React\n});\n"
   ],
   "names":[
      "window",
      "assign",
      "React"
   ],
   "mappings":"AAAA,YAAY,CAAC;AACiB,IAAA,KAAM,WAAN,MAAM,CAAA;AACpCA,KAAM,QAACC,MAAM,CAAC;ICFd,ADGIC,KAAK,EAAEA,KAAK,MAAZA;CACH,CAAC,CAAC"
}
```

This unfortunately, is not fine. 

One thing I notice is that the `import` name is missing. I do not know enough about source maps to know whether it should be present.

**Comment by djalbat** [comment]
I have done some further research and the problem is that SWC is returning different results when called from Watchful compared to when run standalone. Please see my detailed comments on the following pull request:

https://github.com/djalbat/easy-richtextarea/pull/7

I have tried both your `transform()` and `transform/File()` functions, and both return faulty source maps.

I have no idea at this point why SWC should behave differently. I would have assumed your own CLI tool would call the above functions, too.

Over to you.

**Comment by RiESAEX** [comment]
I add a `println!("in rust: {:#?}", result);` at https://github.com/swc-project/swc/blob/32c2de1181c96c873ccb66ac2917fa22409afd11/crates/node/src/transform.rs#L86-L88
build it localy,and link into that repo.
and looks like is not watchful-cli 's fault, it's already broken in rust side.
```
TransformOutput {
    code: "\"use strict\";\nvar _easy = require(\"easy\");\n_easy.window.assign({\n    React: _easy.React\n});\n",
    map: Some(
        "{\"version\":3,\"sources\":[\"lib/example/preamble.js\",\"lib/browser.js\"],\"sourcesContent\":[\"\\\"use strict\\\";\\n\\nimport { React, window } from \\\"easy\\\";\\n\\nwindow.assign({\\n  React\\n});\\n\",\"\\\"use strict\\\";\\n\\nexport { default as Selection } from \\\"./selection\\\";\\n\\nexport { default as RichTextarea } from \\\"./richTextarea\\\";\\n\"],\"names\":[\"window\",\"assign\",\"React\"],\"mappings\":\"AAAA,YAAY,CAAC;AAEiB,IAAA,KAAM,WAAN,MAAM,CAAA;AAEpCA,KAAM,QAACC,MAAM,CAAC;ICJd,ADKEC,KAAK,EAALA,KAAK,MAALA;CACD,CAAC,CAAC\"}",
    ),
}
```
but If I use nodejs to call `transformFile` directly, it generated right sourcemap.
```js
import { transformFile } from "@swc/core";
transformFile("./src/example/preamble.js",{sourceMaps:true}).then((o)=>{
    console.log(o)
})
```
```
{"version":3,"sources":["src/example/preamble.js"],"sourcesContent":["\\"use strict\\";\\n\\nimport { React, window } from \\"easy\\";\\n\\nwindow.assign({\\n  React\\n});\\n"],"names":["window","assign","React"],"mappings":"AAAA,YAAY,CAAC;AAEiB,IAAA,KAAM,WAAN,MAAM,CAAA;AAEpCA,KAAM,QAACC,MAAM,CAAC;IAJd,AAKEC,KAAK,EAALA,KAAK,MAALA;CACD,CAAC,CAAC"}
```

**Comment by RiESAEX** [comment]
@djalbat 
this line watchful override the sourcemap's sources array. It breaks the sourcemap. since the output has two sources.
`sources":["lib/example/preamble.js","lib/browser.js"]`
https://github.com/djalbat/watchful-cli/blob/180d922984d2b64c337397e555716553746a1f96/bin/utilities/transpile.js#L113
after I remove the https://github.com/djalbat/watchful-cli/blob/180d922984d2b64c337397e555716553746a1f96/bin/utilities/transpile.js#L123-L126
the output can be load by the analyzer
![image](https://user-images.githubusercontent.com/89718456/160226114-07ae136b-b428-4d57-8e65-1e4d3004d2c4.png)

**Comment by djalbat** [comment]
Okay, great. I will pick this up Monday evening. Thank you for your efforts.

**Comment by RiESAEX** [comment]
I think I understand why swc output `sources:["lib/example/preamble.js","lib/browser.js"]`, but I'm not sure how to fix.
the commonjs pass make `{React: React}` to `{React: _easy.React}`
like 
```
KeyValueProp {
    key: Ident(
        Ident {
            span: Span {
                lo: BytePos(
                    269,
                ),
                hi: BytePos(
                    274,
                ),
                ctxt: #11,
            },
            sym: Atom('React' type=static),
            optional: false,
        },
    ),
    value: Member(
        MemberExpr {  // the generated member expr has a dummy span
            span: Span {
                lo: BytePos(
                    0,
                ),
                hi: BytePos(
                    0,
                ),
                ctxt: #0,
            },
            obj: Ident(
                Ident {
                    span: Span {
                        lo: BytePos(
                            269,
                        ),
                        hi: BytePos(
                            274,
                        ),
                        ctxt: #12,
                    },
                    sym: Atom('_easy' type=inline),
                    optional: false,
                },
            ),
            prop: Ident(
                Ident {
                    span: Span {
                        lo: BytePos(
                            0,
                        ),
                        hi: BytePos(
                            0,
                        ),
                        ctxt: #0,
                    },
                    sym: Atom('React' type=static),
                    optional: false,
                },
            ),
        },
    ),
}
```
in codegen pass here
https://github.com/swc-project/swc/blob/3761ba4c98d4d9ce488767b4f8abea6bfd9b958d/crates/swc_ecma_codegen/src/lib.rs#L1827-L1830
srcmap! read the span of KeyValueProp, 
KeyValueProp's span should be like `{ lo: key.span.lo, hi: value.span.hi }`, but in this case, value has a dummy span, and Span::new will always make the smaller BytePos be the lo.
https://github.com/swc-project/swc/blob/3761ba4c98d4d9ce488767b4f8abea6bfd9b958d/crates/swc_common/src/syntax_pos.rs#L209-L215
so it get
```
span: Span {
    lo: BytePos(
        0,
    ),
    hi: BytePos(
        269,
    ),
    ctxt: #11,
}
```

then when swc generate the sourcemap, it try to use BytePos(0) to find the sourcefile, and since preamble.js is not the first file, it resolves to browser.js.
https://github.com/swc-project/swc/blob/3761ba4c98d4d9ce488767b4f8abea6bfd9b958d/crates/swc_common/src/source_map.rs#L1178-L1182

**Comment by djalbat** [comment]
Thank you again for your time on this. I have not got time to reply in full now but I think it is clear what the problem is...

I assumed that the 'sources` array would only have one entry, and therefore when I overwrite a `sources` array with three entries with a replacement `sources` array with just one, it is perhaps no surprise that problems occur.

Unfortunately I have to amend the `sources` array in order to get both bundlers and Node to find the original source file. Currently the sources provided by SWC are incorrect. They should be relative paths to the original source file, not paths to the target `lib` file. For example, for a source file of `src/browser.js` and targetFile of `lib/browser.js`, SWC's `sources` array looks like this:

```
[
  "lib/constants.js"
]
```

However it should look like this:

```
[
  "../src/constants.js"
]
```

When I correct to relative paths for all the elements of the `source` array and not just assume that this array contains only one element, I think it is likely that the problem will go away.

By the way, it is good to see the `sourceContents` array being populated again. This was dropped without warning last year. I raised an issue at the time and an option was added to put it back, however it again seems to be the default. So I am happy to take out overwriting of this array.

I won't have time to attend to this fully until Monday evening UK time, now, after which I will post an update and hopefully this bug will go away.
